### ISSUE TITLE
Cannot resolve model_id on ChatHuggingFace, depending on the order of type hints

### ISSUE THREAD

**Issue by aymeric-roucher** [langchain-ai/langchain#17780]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.

### Example Code

from langchain_community.llms import HuggingFaceEndpoint
from langchain.chat_models import ChatHuggingFace

llm = HuggingFaceEndpoint(
    repo_id="HuggingFaceH4/zephyr-7b-beta",
)
agent = ChatHuggingFace(llm=llm)

### Error Message and Stack Trace (if applicable)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb) Cell 11 line 9
      [2](vscode-notebook-cell://ssh-remote%2Bec2/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1) from langchain.chat_models import ChatHuggingFace
      [4](vscode-notebook-cell://ssh-remote%2Bec2/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=3) llm = HuggingFaceEndpoint(
      [5](vscode-notebook-cell://ssh-remote%2Bec2/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=4)     repo_id="HuggingFaceH4/zephyr-7b-beta",
      [6](vscode-notebook-cell://ssh-remote%2Bec2/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=5)     max_new_tokens=100,
      [7](vscode-notebook-cell://ssh-remote%2Bec2/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=6) )
----> [9](vscode-notebook-cell://ssh-remote%2Bec2/home/ubuntu/benchmark_agents/test_prompt_optimizer.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=8) agent = ChatHuggingFace(llm=llm)

File [~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:55](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:55), in ChatHuggingFace.__init__(self, **kwargs)
     [51](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:51) super().__init__(**kwargs)
     [53](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:53) from transformers import AutoTokenizer
---> [55](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:55) self._resolve_model_id()
     [57](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:57) self.tokenizer = (
     [58](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:58)     AutoTokenizer.from_pretrained(self.model_id)
     [59](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:59)     if self.tokenizer is None
     [60](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:60)     else self.tokenizer
     [61](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:61) )

File [~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:155](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:155), in ChatHuggingFace._resolve_model_id(self)
    [152](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:152)         self.model_id = endpoint.repository
    [154](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:154) if not self.model_id:
--> [155](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:155)     raise ValueError(
    [156](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:156)         "Failed to resolve model_id:"
    [157](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:157)         f"Could not find model id for inference server: {endpoint_url}"
    [158](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:158)         "Make sure that your Hugging Face token has access to the endpoint."
    [159](https://vscode-remote+ssh-002dremote-002bec2.vscode-resource.vscode-cdn.net/home/ubuntu/benchmark_agents/~/langchain_fuse_hf_endpoints/libs/community/langchain_community/chat_models/huggingface.py:159)     )

ValueError: Failed to resolve model_id:Could not find model id for inference server: Make sure that your Hugging Face token has access to the endpoint.

### Description

The `model_id` cannot be resolved in [_resolve_model_id](https://github.com/langchain-ai/langchain/blob/865cabff052fe74996bef45faaf00df6f322c215/libs/community/langchain_community/chat_models/huggingface.py#L134), due to the `self.llm` attribute of the ChatHuggingFace object incorrectly being identified as a `HuggingFaceTextGenInference`.

But if we switch the order of [this type hint](https://github.com/langchain-ai/langchain/blob/865cabff052fe74996bef45faaf00df6f322c215/libs/community/langchain_community/chat_models/huggingface.py#L45) to `Union[HuggingFaceEndpoint, HuggingFaceTextGenInference, HuggingFaceHub]`, the llm type is correctly detected again. This seems really fucked up.

### System Info

langchain==0.1.8
langchain-benchmarks==0.0.10
langchain-community==0.0.21
langchain-core==0.1.24
langchainhub==0.1.14

**Comment by aymeric-roucher** [comment]
For now my workaround is to simply replace the line with type hint `llm: HuggingFaceEndpoint`, which works for most usecases.

**Comment by MossabDiae** [comment]
> For now my workaround is to simply replace the line with type hint `llm: HuggingFaceEndpoint`. This could be a general fix since I think few applications have been built on using `ChatHuggingFace` with `HuggingFaceHub` or `HuggingFaceTextGenInference`: but it would be even better to understand why the bug happens in the first place.

Can confirm that this workaround fixes the bug and can correctly generate a `ChatHuggingFace` instance using either `HuggingFaceHub` or `HuggingFaceEndpoint`

```
langchain==0.1.8
langchain-community==0.0.21
langchain-core==0.1.24
```

**Comment by mbanerjeepalmer** [comment]
> For now my workaround is to simply replace the line with type hint `llm: HuggingFaceEndpoint`. This could be a general fix since I think few applications have been built on using `ChatHuggingFace` with `HuggingFaceHub` or `HuggingFaceTextGenInference`: but it would be even better to understand why the bug happens in the first place.

It looks like it's a Pydantic issue, no? It's subclassed from BaseModel and in 1.x versions of Pydantic it uses `left_to_right`. In 2.x it uses `smart`.

**Comment by aymeric-roucher** [comment]
@mbanerjeepalmer it seems to be a Pydantic issue indeed: 
Then do you know a way to resolve the isue and still keep allowing to instantiate the `ChatHuggingFace` from any of these LLMs (`HuggingFaceEndpoint, HuggingFaceHub, HuggingFaceTextGenInference`)?

**Comment by mbanerjeepalmer** [comment]
I don't know too much about Pydantic. My guess is there are two options
1. Change ChatHuggingFace so that the `llm` attribute is typed using `smart_mode`. I haven't been able to get this to work. But I imagine someone smarter than me can.
2. Change the BaseModel so that it uses smart_union in its config. (Probably breaks other things.)

**(1) ChatHuggingFace**
- I'm probably doing something stupid, I don't know why this isn't working: `llm: Annotated[Union[HuggingFaceTextGenInference, HuggingFaceEndpoint, HuggingFaceHub], Field(union_mode="smart")]`
- Pydantic docs: https://docs.pydantic.dev/1.10/usage/schema/#typingannotated-fields

**(2) BaseModel:**
- Config https://docs.pydantic.dev/1.10/usage/model_config/#smart-union
- Where it's probably set https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/language_models/chat_models.py#L130

**Comment by aymeric-roucher** [comment]
Closed by [PR #18693](https://github.com/langchain-ai/langchain/pull/18693)
